src/analysis/normed_space/normed_group_hom.lean

```diff
@@ -533,4 +533,96 @@ lemma norm_noninc_of_isometry (hf : isometry f) : f.norm_noninc :=
 
 end isometry
 
+variables {V₄ W₁ W₂ W₃ : Type*} [semi_normed_group V₄] [semi_normed_group W₁] [semi_normed_group W₂]
+  [semi_normed_group W₃]
+variables (f) (g : normed_group_hom V W)
+variables {f₁ g₁ : normed_group_hom V₁ W₁}
+variables {f₂ g₂ : normed_group_hom V₂ W₂}
+variables {f₃ g₃ : normed_group_hom V₃ W₃}
+
+/-- The equalizer of two morphisms `f g : normed_group_hom V W`. -/
+def equalizer := (f - g).ker
+
+namespace equalizer
+
+/-- The inclusion of `f.equalizer g` as a `normed_group_hom`. -/
+def ι : normed_group_hom (f.equalizer g) V := incl _
+
+lemma comp_ι_eq : f.comp (ι f g) = g.comp (ι f g) :=
+by { ext, rw [comp_apply, comp_apply, ← sub_eq_zero, ← normed_group_hom.sub_apply], exact x.2 }
+
+variables {f g}
+
+/-- If `φ : normed_group_hom V₁ V` is such that `f.comp φ = g.comp φ`, the induced morphism
+`normed_group_hom V₁ (f.equalizer g)`. -/
+@[simps]
+def lift (φ : normed_group_hom V₁ V) (h : f.comp φ = g.comp φ) :
+  normed_group_hom V₁ (f.equalizer g) :=
+{ to_fun := λ v, ⟨φ v, show (f - g) (φ v) = 0,
+    by rw [normed_group_hom.sub_apply, sub_eq_zero, ← comp_apply, h, comp_apply]⟩,
+  map_add' := λ v₁ v₂, by { ext, simp only [map_add, add_subgroup.coe_add, subtype.coe_mk] },
+  bound' := by { obtain ⟨C, C_pos, hC⟩ := φ.bound, exact ⟨C, hC⟩ } }
+
+@[simp] lemma lift_ι (φ : normed_group_hom V₁ V) (h : f.comp φ = g.comp φ) :
+  (ι _ _).comp (lift φ h) = φ :=
+by { ext, refl }
+
+@[simps]
+def lift_equiv : {φ : normed_group_hom V₁ V // f.comp φ = g.comp φ} ≃
+  normed_group_hom V₁ (f.equalizer g) :=
+{ to_fun := λ φ, lift φ φ.prop,
+  inv_fun := λ ψ, ⟨(ι f g).comp ψ, by { rw [← comp_assoc, ← comp_assoc, comp_ι_eq] }⟩,
+  left_inv := λ φ, by simp,
+  right_inv := λ ψ, by { ext, refl } }
+
+/-- Given `φ : normed_group_hom V₁ V₂` and `ψ : normed_group_hom W₁ W₂` such that
+`ψ.comp f₁ = f₂.comp φ` and `ψ.comp g₁ = g₂.comp φ`, the induced morphism
+`normed_group_hom (f₁.equalizer g₁) (f₂.equalizer g₂)`. -/
+def map (φ : normed_group_hom V₁ V₂) (ψ : normed_group_hom W₁ W₂)
+  (hf : ψ.comp f₁ = f₂.comp φ) (hg : ψ.comp g₁ = g₂.comp φ) :
+  normed_group_hom (f₁.equalizer g₁) (f₂.equalizer g₂) :=
+lift (φ.comp $ ι _ _) $
+by { simp only [← comp_assoc, ← hf, ← hg], simp only [comp_assoc, comp_ι_eq] }
+
+variables {φ : normed_group_hom V₁ V₂} {ψ : normed_group_hom W₁ W₂}
+variables {φ' : normed_group_hom V₂ V₃} {ψ' : normed_group_hom W₂ W₃}
+
+@[simp] lemma map_ι (hf : ψ.comp f₁ = f₂.comp φ) (hg : ψ.comp g₁ = g₂.comp φ) :
+  (ι f₂ g₂).comp (map φ ψ hf hg) = φ.comp (ι _ _) :=
+lift_ι _ _
+
+@[simp] lemma map_id : @map _ _ _ _ _ _ _ _ f₁ g₁ f₁ g₁ (id V₁) (id W₁) rfl rfl = id _ :=
+by { ext, refl }
```

Comment:
Can you write this without the `@`s?

Reply:
Something like
```lean
@[simp] lemma map_id : map (id V₁) (id W₁) (show (id W₁).comp f₁ = f₁.comp (id V₁), from rfl)
  (show (id W₁).comp g₁ = g₁.comp (id V₁), from rfl) = id _ :=
by { ext, refl }
```
works, but I am not sure it's better than the `@`...

Reply:
Maybe:
```
@[simp] lemma map_id : map (id V₁) (id W₁) (rfl : (id W₁).comp f₁ = f₁.comp (id V₁))
  (rfl : (id W₁).comp g₁ = g₁.comp (id V₁)) = id _ :=
by { ext, refl }
```
Which is a bit shorter? Generally a bit of extra verbosity is still worth avoiding `@` for, as it isn't fragile to extra typeclass arguments being added

Reply:
Alternatively, can you fill the underscore on the RHS to avoid needing the ascriptionson the rfls?

Reply:
Ah you're right! Filling the `_` on the RHS solves the problem.